tensorflow/core/transforms/remapper/pass.cc

```diff
@@ -306,6 +312,241 @@ class BasePatternActivationRewriter : public BasePatternRewriter {
+      : RemapperPatternBase(op_name, helper, benefit) {}
+
+  using Pattern = FusedBatchNormEx;
+
+  bool is_valid_batch_norm(Operation *fused_batch_norm_op) const {
+    TFOp fusedbatchnormop_wrapper(fused_batch_norm_op);
+    if (!this->helper_.getDialect()->IsFusedBatchNorm(fusedbatchnormop_wrapper))
+      return false;
+    // We fuse FusedBatchNorm on GPU or oneDNN CPU.
+    if (!this->helper_.isOneDNNEnabled() &&
+        !util::OpHasDevice(fused_batch_norm_op, tensorflow::DEVICE_GPU))
+      return false;
+
+    TypeAttr attr = fused_batch_norm_op->getAttrOfType<TypeAttr>("T");
+    if (!attr) return false;
+    Type dtype_T = attr.getValue();
+
+    if (util::OpHasDevice(fused_batch_norm_op, tensorflow::DEVICE_GPU)) {
+      // GPU supports float and half.
+      // Put this condition before check `isOneDNNEnabled()` because this node
+      // should be processed when it's on GPU and oneDNN CPU is enabled.
+      if (!dtype_T.isa<Float32Type, Float16Type>()) return false;
+    } else {
+      // Bfloat16 is available only with oneDNN.
+      // Half is not available with oneDNN.
+      if (this->helper_.isOneDNNEnabled() && !dtype_T.isa<Float32Type>() &&
+          !dtype_T.isa<BFloat16Type>()) {
+        return false;
+      }
+    }
+
+    // Get the FusedBatchNorm training mode.
+    auto training_attr =
+        fused_batch_norm_op->getAttrOfType<BoolAttr>("is_training");
+    if (!training_attr) return false;
+    auto is_training = training_attr.getValue();
+
+    auto data_format_attr =
+        fused_batch_norm_op->getAttrOfType<StringAttr>("data_format");
+    if (!data_format_attr) return false;
+    auto data_format = data_format_attr.getValue();
+
+    if (data_format != "NHWC" && data_format != "NCHW") return false;
+
+    // In training mode we rely on cuDNN for computing FusedBatchNorm with side
+    // inputs and activation, and it has its own limitations. In inference mode
+    // we have a custom CUDA kernel that doesn't not have these constraints.
+    if (is_training &&
+        util::OpHasDevice(fused_batch_norm_op, tensorflow::DEVICE_GPU)) {
+      // cuDNN only supports NHWC data layout.
+      if (data_format != "NHWC") return false;
+
+      // Data type must be Float16.
+      if (!dtype_T.isa<Float16Type>()) return false;
+
+      // Channel dimension must be a multiple of 4.
+      auto fbn_input0_shape =
+          fused_batch_norm_op->getOperand(0).getType().cast<ShapedType>();
+      auto fbn_input0_shape_dims = fbn_input0_shape.getShape();
+
+      const bool valid_channel_dim = (fbn_input0_shape.getRank() == 4) &&
+                                     (fbn_input0_shape_dims[3] % 4 == 0);
+
+      if (!valid_channel_dim) return false;
+
+      // cuDNN must support CUDNN_BATCHNORM_SPATIAL_PERSISTENT mode.
+      if (!BatchnormSpatialPersistentEnabled()) return false;
+    }
+
+    // FusedBatchNormV2 and V3 have an extra type parameter.
+    if (fused_batch_norm_op->getName().getStringRef().str() !=
```

Comment:
```suggestion
    if (fused_batch_norm_op->getName().getStringRef() !=
```

If this is supposed to be a frequent operation, you may consider to cache the StringAttr for it

Reply:
Changed to suggested.